File: src/main/java/com/github/alexandrecarlton/idea/settings/dagger/configuration/ConfigurationSubcomponent.java
package com.github.alexandrecarlton.idea.settings.dagger.configuration;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.BeforeLaunchConfigurationSettings;
import dagger.BindsInstance;
import dagger.Subcomponent;

import javax.inject.Named;

@Subcomponent(modules = {
  ConfigurationModule.class,
  OptionalSettingsApplierModule.class,
  SettingsApplierModule.class})
public interface ConfigurationSubcomponent {

  SettingsApplier<BeforeLaunchConfigurationSettings> beforeLaunchConfigurationSettingsApplier();

  @Subcomponent.Builder
  interface Builder {
    @BindsInstance
    ConfigurationSubcomponent.Builder configuration(@Named("configuration") String name);
    ConfigurationSubcomponent build();
  }
}

File: dagger/src/main/kotlin/com/github/alexandrecarlton/idea/settings/configuration/ConfigurationSubcomponent.kt
package com.github.alexandrecarlton.idea.settings.configuration

import com.github.alexandrecarlton.idea.settings.dialog.SettingsApplier
import com.github.alexandrecarlton.idea.settings.dialog.configurations.common.before_launch.BeforeLaunchConfigurationSettings
import com.github.alexandrecarlton.idea.settings.dialog.configurations.common.before_launch.BeforeLaunchModule
import dagger.BindsInstance
import dagger.Subcomponent

import javax.inject.Named

@Subcomponent(modules = [
    BeforeLaunchModule::class,
    ConfigurationModule::class])
interface ConfigurationSubcomponent {

    fun beforeLaunchConfigurationSettingsApplier(): SettingsApplier<BeforeLaunchConfigurationSettings>

    @Subcomponent.Builder
    interface Builder {
        @BindsInstance
        fun configuration(@Named("configuration") name: String): Builder

        fun build(): ConfigurationSubcomponent
    }
}
